src/mem/epoch/tagged_atomic.rs

```diff
@@ -0,0 +1,295 @@
+    {
+        guard_tag::<T>(new_tag);
+        if self.ptr.compare_and_swap(
+               opt_shared_into_usize(old, old_tag),
+               new.as_raw() as usize | new_tag,
+               ord
+           ) == opt_shared_into_usize(old, old_tag)
+        {
+            Ok(unsafe { Shared::from_owned(new) })
+        } else {
+            Err(new)
+        }
+    }
+
+    /// Do a compare-and-set on the pair, from a `Shared` to another `Shared` pointer in the pair
+    /// with the given memory ordering.
+    ///
+    /// The boolean return value is `true` when the CAS is successful.
+    ///
+    /// # Panics
+    ///
+    /// Panics if either `old_tag` or `new_tag` is `>= mem::align_of::<T>()`.
+    pub fn cas_shared(&self,
+                      old: Option<Shared<T>>, old_tag: usize,
+                      new: Option<Shared<T>>, new_tag: usize,
+                      ord: Ordering
+                     ) -> bool
+    {
+        self.ptr.compare_and_swap(opt_shared_into_usize(old, old_tag),
+                                  opt_shared_into_usize(new, new_tag),
+                                  ord) == opt_shared_into_usize(old, old_tag)
+    }
+
+    /// Do an atomic swap with an `Owned` pointer in the new pair with the given memory ordering.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `new_tag >= mem::align_of::<T>()`.
+    pub fn swap<'a>(&self, new: Option<Owned<T>>, new_tag: usize, ord: Ordering, _: &'a Guard)
+                    -> (Option<Shared<'a, T>>, usize) {
+        let (ptr, tag) = unpack_tag(self.ptr.swap(opt_owned_into_usize(new, new_tag), ord));
+        unsafe { (Shared::from_raw(ptr), tag) }
+    }
+
+    /// Do an atomic swap with a `Shared` pointer in the new pair with the given memory ordering.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `new_tag >= mem::align_of::<T>()`.
+    pub fn swap_shared<'a>(&self,
+                           new: Option<Shared<T>>, new_tag: usize,
+                           ord: Ordering,
+                           _: &'a Guard
+                          ) -> (Option<Shared<'a, T>>, usize) {
+        let (ptr, tag) = unpack_tag(self.ptr.swap(opt_shared_into_usize(new, new_tag), ord));
+        unsafe { (Shared::from_raw(ptr), tag) }
+    }
+
+    /// Perform a bitwise or on the current tag and the argument `tag` and set the new tag to
+    /// the result. Returns the previous value of the tag.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `tag >= mem::align_of::<T>()`.
+    pub fn or_tag(&self, tag: usize, ord: Ordering) -> usize {
+        guard_tag::<T>(tag);
+        self.ptr.fetch_or(tag, ord)
+    }
+
+    /// Perform a bitwise and on the current tag and the argument `tag` and set the new tag to
+    /// the result. Returns the previous value of the tag.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `tag >= mem::align_of::<T>()`.
+    pub fn and_tag(&self, tag: usize, ord: Ordering) -> usize {
```

Comment:
Added `and_tag` because why not, changed the docs to `bitwise`, since it's only a logical operation for booleans and made the two functions return `usize` in case anybody needs it, if not it'll be optimized out anyway.